File: leetcode/src/test/java/com/giwankim/leetcode/SimplifyPathTest.java
package com.giwankim.leetcode;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

class SimplifyPathTest {
  SimplifyPath sut = new SimplifyPath();

  @ParameterizedTest
  @CsvSource(
      value = {
        "/home/:/home",
        "/home//foo:/home/foo",
        "/home/user/Documents/../Pictures:/home/user/Pictures",
        "/../:/",
        "/a/./b/../../c/:/c"
      },
      delimiter = ':')
  void simplifyPath(String path, String expected) {
    assertThat(sut.simplifyPath(path)).isEqualTo(expected);
  }
}

File: leetcode-kt/src/test/kotlin/com/giwankim/leetcode/SimplifyPathTest.kt
package com.giwankim.leetcode

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe

class SimplifyPathTest :
    FunSpec(
        {
            val sut = SimplifyPath()

            test("root") {
                sut.simplifyPath("/") shouldBe "/"
            }

            context("one depth") {
                test("base case") {
                    sut.simplifyPath("/home") shouldBe "/home"
                }

                test("trailing comma") {
                    sut.simplifyPath("/home/") shouldBe "/home"
                }

                test("cannot go up in depth from root") {
                    sut.simplifyPath("/../") shouldBe "/"
                }
            }

            context("two depth") {
                test("base case") {
                    sut.simplifyPath("/home/foo") shouldBe "/home/foo"
                }

                test("ignore consecutive slashes") {
                    sut.simplifyPath("/home//foo") shouldBe "/home/foo"
                }

                test(".. moves up one depth") {
                    sut.simplifyPath("/home/../foo") shouldBe "/foo"
                }
            }

            context("general case") {
                test("complex path") {
                    sut.simplifyPath("/home/user/Documents/../Pictures") shouldBe "/home/user/Pictures"
                }

                test("... is a valid directory name") {
                    sut.simplifyPath("/.../a/../b/c/../d/./") shouldBe "/.../b/d"
                }
            }
        },
    )
